FibonacciHeap.java

```diff
@@ -449,62 +449,78 @@ public void delete(HeapNode x) {
     * to reflect this change (for example, the cascading cuts procedure should be applied if needed).
     */
     public void decreaseKey(HeapNode x, int delta) {
-        if (this.isEmpty() || x == null) {
+        if (this.isEmpty() || x == null || delta < 0) {
             return;
         }
 
+        HeapNode keyToDecrease = x;
         x.key -= delta;
-        HeapNode parent = x.parent;
+        int decreasedValue = x.key;
 
-        if (parent != null && x.key < parent.key) {
-            cutNode(x, parent);
-            cascadeCut(parent);
+        if (x.getParent() == null) {
+            if (x.getKey() < this.min.getKey())
+                this.min = x;
+            return;
         }
 
-        updateMin(x);
-    }
+        if (x.getParent().getKey() <= x.getKey()) {
+            return;
+        }
+
+        x.mark = true;
+
+        //if we got here, we need to cut x.
+        while(x.isMarked()) {
+            HeapNode parent = x.parent;
+            cutNode(x);
 
-    private void cutNode(HeapNode x, HeapNode parent) {
-        if (x == x.next) {
-            parent.child = null;
+            if (x.isMarked()) {
+                x.mark = false;
+                this.markedCount--;
+            }
+
+            x = parent;
         }
 
-        x.prev.next = x.getNext();
-        x.next.prev = x.getPrev();
-        if (x == parent.child) {
-            x.child = parent.getNext();
+        // update min.
+        if (decreasedValue < this.min.getKey()) {
+            min = keyToDecrease;
         }
 
-        x.rank = parent.getRank() - 1;
-        x.next = x;
-        x.prev = x;
+        if (x.getParent() != null) {
+            x.mark = true;
+            this.markedCount++;
+        }
+    }
 
-        this.min.prev.next = x;
-        x.next = this.getMin();
-        x.prev = this.getMin().getNext();
+    public void cutNode(HeapNode x) {
```

Comment:
At each node you cut you need to mark the parent
just add
`x.parent.mark()`
`this.markedCount++`